litgpt/pretrain.py

```diff
@@ -392,26 +498,91 @@ def get_lr(learning_rate: float, it: int, warmup_iters: int, max_iters: int, min
     return min_lr + coeff * (learning_rate - min_lr)
 
 
-def initialize_weights(fabric: L.Fabric, model: GPT, n_layer: int, n_embd: int) -> None:
+def initialize_weights(fabric: L.Fabric, model: GPT, n_layer: int, n_embd: int, fast_init: bool) -> None:
     """GPT-NeoX weight initialization (https://arxiv.org/abs/2204.06745)."""
     # Adapted from https://github.com/jzhang38/TinyLlama
 
     def init_weights(module, std):
         nn.init.normal_(module.weight, mean=0.0, std=std)
         if getattr(module, "bias", None) is not None:
             nn.init.zeros_(module.bias)
+    
+    fabric.print(f"Initializing weights with {fast_init=}.", flush=True)
+    if fast_init:
```

Comment:
How did you test fast_init?

Reply:
I ported over the same functionality from the old redrock_trainer code and then ran with fast_init off and on. I can re-run both and post init times here to compare them.

Reply:
That would be great thanks. I just want to make sure this is doing what we expect.

Reply:
Just finished running this. It looks like the default initialization is faster than fast_init, however the total benchmark runtime is significantly faster with fast_init=True. 

**fast_init=True:**

Total Benchmark Run Time: 0:09:16.492884 

```
I0530 23:38:40.395345     267 entrypoint_utils.py:189] Initializing weights with fast_init=True.
I0530 23:38:47.845298     267 entrypoint_utils.py:189] 0 time to init weights: 7.45s
I0530 23:38:47.902600     267 entrypoint_utils.py:189] Time to instantiate model: 7.47 seconds.
I0530 23:38:47.902718     267 entrypoint_utils.py:189] Total parameters: 6,857,302,016
I0530 23:38:49.181852     267 entrypoint_utils.py:189] /usr/local/lib/python3.10/dist-packages/lightning/fabric/fabric.py:954: The model passed to `Fabric.setup()` has 130 parameters on different devices (for example 'transformer.h.0.norm_1.weight' on meta and 'lm_head.weight' on cpu). Since `move_to_device=True`, all parameters will be moved to the new device. If this is not desired, set `Fabric.setup(..., move_to_device=False)`.
```

**fast_init=False:**

Total Benchmark Run Time: 0:14:16.137701

```
I0531 00:10:51.352488     266 entrypoint_utils.py:189] Initializing weights with fast_init=False.
I0531 00:10:51.352526     266 entrypoint_utils.py:189] 0 time to init weights: 0.00s
I0531 00:10:51.408955     266 entrypoint_utils.py:189] Time to instantiate model: 0.02 seconds.
I0531 00:10:51.409012     266 entrypoint_utils.py:189] Total parameters: 6,857,302,016
```

Reply:
Probably we need to test with a larger model (something like Llama 70B). I'm fine to check this in though.

Reply:
Awesome, merging now.